File: workshop/example/python/parser/character.py
class Character:
    def __init__(self, target):
        self.target = target
    
    def parse(self, input):
        if input[0] == self.target:
            return [(str(self.target), input[1:])]
        else:
            return []

def character(target):
    return Character(target)

if __name__ == '__main__':
    assert character('A').parse('ABC') == [('A', 'BC')]
    assert character('A').parse('BC') == []
    assert character('A').parse('aBC') == []
    assert character('a').parse('aBC') == [('a', 'BC')]

File: workshop/example/java/src/main/java/nl/fifthpostulate/parser/base/Character.java
package nl.fifthpostulate.parser.base;

import nl.fifthpostulate.parser.Pair;
import nl.fifthpostulate.parser.Parser;

import java.util.Collections;
import java.util.List;

import static nl.fifthpostulate.parser.Pair.pair;
import static nl.fifthpostulate.parser.base.Predicate.predicate;

public record Character(java.lang.Character target) implements Parser<java.lang.Character> {
    @Override
    public List<Pair<java.lang.Character, String>> parse(String input) {
        if (input.startsWith(String.valueOf(target))) {
            return Collections.singletonList(pair(target, input.substring(1)));
        } else {
            return Collections.emptyList();
        }
    }

    public static Parser<java.lang.Character> character(char target) {
        return predicate(c -> c == target);
    }
}
